src/partr.c

```diff
@@ -0,0 +1,1149 @@
+// This file is a part of Julia. License is MIT: https://julialang.org/license
+
+#include <assert.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <strings.h>
+
+#include "julia.h"
+#include "julia_internal.h"
+#include "gc.h"
+#include "threading.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#ifdef JULIA_ENABLE_THREADING
+#ifdef JULIA_ENABLE_PARTR
+
+// empirically, finish_task needs about 64k stack space to infer/run
+// and additionally, gc-stack reserves 64k for the guard pages
+#if defined(MINSIGSTKSZ) && MINSIGSTKSZ > 131072
+#define MINSTKSZ MINSIGSTKSZ
+#else
+#define MINSTKSZ 131072
+#endif
+
+// task states and stack switching
+extern jl_sym_t *done_sym;
+extern jl_sym_t *failed_sym;
+extern jl_sym_t *runnable_sym;
+extern void jl_switchto(jl_task_t **pt);
+extern char *jl_alloc_fiber(jl_ucontext_t *t, size_t *ssize, jl_task_t *owner);
+
+// the lovely task-done-hook hack
+extern jl_function_t *task_done_hook_func;
+
+// GC functions used
+extern int jl_gc_mark_queue_obj_explicit(jl_gc_mark_cache_t *gc_cache,
+                                         jl_gc_mark_sp_t *sp, jl_value_t *obj);
+
+// thread sleep threshold
+extern uint64_t jl_thread_sleep_threshold;
+
+// multiq
+// ---
+
+/* a task heap */
+typedef struct taskheap_tag {
+    jl_mutex_t lock;
+    jl_task_t **tasks;
+    int16_t ntasks, prio;
+} taskheap_t;
+
+/* multiqueue parameters */
+static const int16_t heap_d = 8;
+static const int heap_c = 4;
+
+/* size of each heap */
+static const int tasks_per_heap = 8192; // TODO: this should be smaller by default, but growable!
+
+/* the multiqueue's heaps */
+static taskheap_t *heaps;
+static int16_t heap_p;
+
+/* unbias state for the RNG */
+static uint64_t cong_unbias;
+
+/* for thread sleeping */
+static uv_mutex_t sleep_lock;
+static uv_cond_t  sleep_alarm;
+
+
+/*  multiq_init()
+ */
+static inline void multiq_init(void)
+{
+    heap_p = heap_c * jl_n_threads;
+    heaps = (taskheap_t *)calloc(heap_p, sizeof(taskheap_t));
+    for (int16_t i = 0;  i < heap_p;  ++i) {
```

Comment:
Why the `int16_t` for the iterator here? Perhaps a native width `int` (or `size_t`) would make more sense?

Same comment applies to various iterators within partr.c - some are `int16_t`, some `int`, there's even an `int64_t`.

Reply:
Right. I think it should mainly be necessary to match signed vs unsigned in comparisons (not so much bit width) to avoid surprises and related warnings arising from the likes of

```C
#include <stdio.h>
int main() {
    if (-1 > (size_t)0)
        printf("Surprise!\n");
    return 0;
}
```

I mainly noticed this because I was having a read through for where 32 and 64 bit versions might differ.